lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1314 @@
+		FROM MsgReceivedChange, TO MsgSentChange,
+		FROM ReceivedTxComplete, TO MsgSentChange
+	]);
+	define_state_machine_transitions!(received_tx_add_output, &msgs::TxAddOutput, [
+		FROM MsgSentChange, TO MsgReceivedChange,
+		FROM SentTxComplete, TO MsgReceivedChange
+	]);
+
+	// TxRemoveInput
+	define_state_machine_transitions!(sent_tx_remove_input, &msgs::TxRemoveInput, [
+		FROM MsgReceivedChange, TO MsgSentChange,
+		FROM ReceivedTxComplete, TO MsgSentChange
+	]);
+	define_state_machine_transitions!(received_tx_remove_input, &msgs::TxRemoveInput, [
+		FROM MsgSentChange, TO MsgReceivedChange,
+		FROM SentTxComplete, TO MsgReceivedChange
+	]);
+
+	// TxRemoveOutput
+	define_state_machine_transitions!(sent_tx_remove_output, &msgs::TxRemoveOutput, [
+		FROM MsgReceivedChange, TO MsgSentChange,
+		FROM ReceivedTxComplete, TO MsgSentChange
+	]);
+	define_state_machine_transitions!(received_tx_remove_output, &msgs::TxRemoveOutput, [
+		FROM MsgSentChange, TO MsgReceivedChange,
+		FROM SentTxComplete, TO MsgReceivedChange
+	]);
+
+	// TxComplete
+	define_state_machine_transitions!(sent_tx_complete, &msgs::TxComplete, [
+		FROM MsgReceivedChange, TO SentTxComplete,
+		FROM ReceivedTxComplete, TO NegotiationComplete
+	]);
+	define_state_machine_transitions!(received_tx_complete, &msgs::TxComplete, [
+		FROM MsgSentChange, TO ReceivedTxComplete,
+		FROM SentTxComplete, TO NegotiationComplete
+	]);
+}
+
+pub struct InteractiveTxConstructor {
+	state_machine: StateMachine,
+	channel_id: ChannelId,
+	inputs_to_contribute: Vec<(SerialId, TxIn, TransactionU16LenLimited)>,
+	outputs_to_contribute: Vec<(SerialId, TxOut)>,
+}
+
+pub enum InteractiveTxMessageSend {
+	TxAddInput(msgs::TxAddInput),
+	TxAddOutput(msgs::TxAddOutput),
+	TxComplete(msgs::TxComplete),
+}
+
+// This macro executes a state machine transition based on a provided action.
+macro_rules! do_state_transition {
+	($self: ident, $transition: ident, $msg: expr) => {{
+		let state_machine = core::mem::take(&mut $self.state_machine);
+		$self.state_machine = state_machine.$transition($msg);
+		match &$self.state_machine {
+			StateMachine::NegotiationAborted(state) => Err(state.0.clone()),
+			_ => Ok(()),
+		}
+	}};
+}
+
+fn generate_holder_serial_id<ES: Deref>(entropy_source: &ES, is_initiator: bool) -> SerialId
+where
+	ES::Target: EntropySource,
+{
+	let rand_bytes = entropy_source.get_secure_random_bytes();
+	let mut serial_id_bytes = [0u8; 8];
+	serial_id_bytes.copy_from_slice(&rand_bytes[..8]);
+	let mut serial_id = u64::from_be_bytes(serial_id_bytes);
+	if serial_id.is_for_initiator() != is_initiator {
+		serial_id ^= 1;
+	}
+	serial_id
+}
+
+impl InteractiveTxConstructor {
+	pub fn new<ES: Deref>(
+		entropy_source: &ES, channel_id: ChannelId, feerate_sat_per_kw: u32, is_initiator: bool,
+		tx_locktime: AbsoluteLockTime, inputs_to_contribute: Vec<(TxIn, TransactionU16LenLimited)>,
+		outputs_to_contribute: Vec<TxOut>, to_remote_value: u64,
+	) -> (Self, Option<InteractiveTxMessageSend>)
```

Comment:
While this is internal code, it would be nice to have some documentation including what the return values are for.